keras_core/optimizers/base_optimizer.py

```diff
@@ -321,6 +327,22 @@ def stateless_apply(self, optimizer_variables, grads, trainable_variables):
                 optimizer_variables.append(v)
         return trainable_variables, optimizer_variables
 
+    def scale_loss(self, loss):
+        """Scale the loss before computing gradients.
+
+        Scales the loss before gradients are computed in a `train_step`. This
+        is primarily useful during mixed precision training to prevent numeric
+        underflow.
+        """
+        if self.loss_scale_factor is not None:
+            return loss * self.loss_scale_factor
+        return loss
+
+    def stateless_scale_loss(self, optimizer_variables, loss):
```

Comment:
Good idea. I like the stateless scope. That keeps the weirdness isolated to jax.

Overall the jax trainer is definitely a little wonky (pushing state through `aux` for `jax.value_and_grad` is kinda bleh), but the real grossness is the jax implementation of the loss scale optimizer itself. Basically all jax control flow cannot be written then same was as tf/torch because a stateless scope is insufficient, you still have to return all state from each control flow callback. Not something to fix this PR, and I'm not sure we can really do anything about it, but it does break the idea of a backend agnostic graph of `ops`.